### PR TITLE
Add an option to _search to force synthetic source

### PR DESCRIPTION
**Pull request by nik9000**
This adds `?force_synthetic_source` to, well, force running the fetch
phase with synthetic source. If the mapping is incompatible with
synthetic source it'll throw a 400 error. This should be useful for
folks looking to evaluate synthetic source.

Folks can use it to check if their current mapping supports synthetic
source. Given how picky synthetic source is, their mapping probably
*doesn't* support it now, but this gives us a very quick way to test
it. Which should help shorten the testing cycle.

If their mapping does support synthetic source it'll fetch the _source
as thought synthetic source were enabled so folks can have a look at
the result and see if it's done a good job. Synthetic source makes a
lot of assumptions about the original layout of the _source and some
of those assumptions are destructive to, well, the layout. If you
need the layout you can try to use the results this produces.

Finally this gives you an upper bound on the performance hit of synthetic
source. This feature will *still* load and decompress the original
_source bytes, so you won't get any of the fetch-time performance
advantages to synthetic source - but it throws out the _source and
rebuilds it from doc values - do you get all of the fetch time
performance disadvanteages. It the fethc performance is still
acceptable then you know synthetic source is ok for you. If it is
marginal then you'd need a full scale test, reindexing everything
with synthetic source. If the fetch performance is nowhere near ok
then you know synthetic source isn't good for you, at least not yet.

### CODE REVIEW

**Comment by nik9000** [comment]
This actually won't "turn off" the _source - it'll still be decompressed. So this will show a "worst case" for latency - it'll include the synthetic source. It'll just load the `_source` from stored fields and throw it away.

**Comment by javanna** [comment]
Heya @nik9000 can you expand on what the usecase is for this feature?

**Comment by nik9000** [comment]
> Heya @nik9000 can you expand on what the usecase is for this feature?

I've edited the description to add it - the short version is that it's a way to evaluate synthetic source without reindexing. I don't think it's a "forever" thing, but I've talked to quite a few folks that would find it useful.

**Comment by sethmlarson** [comment]
> I don't think it's a "forever" thing

Is this option going to be deprecated+removed in the future? If so might be good to document it this way. Also is this option for human debugging only or are we planning on users using it in applications?

**Comment by nik9000** [comment]
> Is this option going to be deprecated+removed in the future?

I think so.

> Also is this option for human debugging only or are we planning on users using it in applications?

Debugging. Though maybe not just humans.

**Comment by sethmlarson** [comment]
@nik9000 Gotcha! In that case can we document that this parameter is experimental and will be removed in a future version? I think that's enough warning for a parameter that's behind a feature flag and undocumented elsewhere.

File: server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java [hunk 1]
```diff
@@ -208,6 +209,9 @@ public static void parseSearchRequest(
                 request.paramAsBoolean("ccs_minimize_roundtrips", searchRequest.isCcsMinimizeRoundtrips())
             );
         }
+        if (IndexSettings.isTimeSeriesModeEnabled() && request.paramAsBoolean("force_synthetic_source", false)) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java:212]: Just to check: `IndexSettings.isTimeSeriesModeEnabled()` is the feature flag, yes?

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 2]
```diff
@@ -300,13 +320,15 @@ public ShardSearchRequest(ShardSearchRequest clone) {
...
         this.waitForCheckpointsTimeout = clone.waitForCheckpointsTimeout;
+        this.forceSyntheticSource = clone.forceSyntheticSource;
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         innerWriteTo(out, false);
         OriginalIndices.writeOriginalIndices(originalIndices, out);
+
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:331]: nit: no new line here

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 3]
```diff
@@ -357,6 +379,13 @@ protected final void innerWriteTo(StreamOutput out, boolean asKey) throws IOExce
                     + "] or greater."
             );
         }
+        if (out.getVersion().onOrAfter(Version.V_8_4_0)) {
+            out.writeBoolean(forceSyntheticSource);
+        } else {
+            if (forceSyntheticSource) {
+                throw new IllegalArgumentException("force_synethic_source is not supported before 8.3.0");
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:386]: nit: spelling `force_synethic_source`

**Review by romseygeek — COMMENTED** [review]
Looks good in general.  I wonder if it would make more sense to make this flag state on SearchExecutionContext and then choose whether or not to build a synthetic source in SEC.getSourceLoader(), rather than pushing this down into MappingLookup though?

**Comment by nik9000** [comment]
> I wonder if it would make more sense to make this flag state on SearchExecutionContext and then choose whether or not to build a synthetic source in SEC.getSourceLoader(), rather than pushing this down into MappingLookup though?

`SearchExecutionContext` can't see the `root` object mapper that's required. I don't think we want to make that public for other reasons.

**Comment by romseygeek** [comment]
> SearchExecutionContext can't see the root object mapper that's required. I don't think we want to make that public for other reasons.

It can via `mappingLookup.getMapping().getRoot()`

**Comment by nik9000** [comment]
> It can via `mappingLookup.getMapping().getRoot()`

"'getRoot()' is not public in 'org.elasticsearch.index.mapper.Mapping'. Cannot be accessed from outside package"

**Comment by romseygeek** [comment]
Fair point...

But! SourceFieldMapper is in the same package, and we can change the signature of `SourceFieldMapper.newSourceLoader()` to take the Mapping instead of the root object; and indeed, everywhere that `newSourceLoader()` is called, it's called with a call to `getMapping().getRoot()`.  So if we rearrange things a bit then we can keep package-privacy and avoid adding that boolean to the mapping APIs.

**Comment by nik9000** [comment]
> But! SourceFieldMapper is in the same package, and we can change the signature of `SourceFieldMapper.newSourceLoader()` to take the Mapping instead of the root object; and indeed, everywhere that `newSourceLoader()` is called, it's called with a call to `getMapping().getRoot()`. So if we rearrange things a bit then we can keep package-privacy and avoid adding that boolean to the mapping APIs.

I'll have a look

**Comment by nik9000** [comment]
> I'll have a look

Yeah, that works. @romseygeek, have another look when  you get a chance.

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @nik9000!

**Comment by nik9000** [comment]
@sethmlarson I've added a flag to the spec to mark this parameter as behind the feature flag. Is that OK? Am I going to break a everyone?

File: rest-api-spec/src/main/resources/rest-api-spec/api/search.json [hunk 4]
```diff
@@ -77,6 +77,10 @@
         "type":"number",
         "description":"Starting offset (default: 0)"
       },
+      "force_synthetic_source":{
+        "type":"boolean",
+        "description":"Should we force synthetic source for all fetches from this index?"
+      },
```
Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/api/search.json]: Woops - dupes. One moment.

**Review by sethmlarson — APPROVED** [review]
The change to the rest-api-spec schema.json LGTM.

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 5]
```diff
@@ -357,6 +381,13 @@ protected final void innerWriteTo(StreamOutput out, boolean asKey) throws IOExce
                     + "] or greater."
             );
         }
+        if (out.getVersion().onOrAfter(Version.V_8_4_0)) {
+            out.writeBoolean(forceSyntheticSource);
+        } else {
+            if (forceSyntheticSource) {
+                throw new IllegalArgumentException("force_synthetic_source is not supported before 8.3.0");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:388]: you mention 8.3 in the error message, but 8.4 in the conditional above. I suspect you did that on purpose but I am not following: is there a way to force synthetic source on 8.3 which does not support the flag?

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:388]: A mistake!

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 6]
```diff
@@ -308,6 +321,9 @@ public void writeTo(StreamOutput out) throws IOException {
                     + "] or greater."
             );
         }
+        if (out.getVersion().onOrAfter(Version.V_8_4_0)) {
+            out.writeBoolean(forceSyntheticSource);
+        }
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:326]: Yikes. Sorry, yeah. I should have had it.

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java (hunk 5 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:388]: Unbelievable, I could not believe my eyes hence I convinced myself you must have done it on purpose :)

**Pull request merged**